src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp

```diff
@@ -0,0 +1,730 @@
+
+        if (osl_param.hasDefault)
+            def_val = static_cast<float>(osl_param.defaultValue.at(0));
+
+        const int ctrl_id_1 = ctrl_id++;
+        const int ctrl_id_2 = ctrl_id++;
+
+        pb_desc->AddParam(
+            param_id++,
+            param_str.c_str(),
+            TYPE_FLOAT,
+            P_ANIMATABLE,
+            string_id,
+            p_default, def_val,
+            p_range, min_val, max_val,
+            p_ui, TYPE_SPINNER, EDITTYPE_FLOAT, ctrl_id_1, ctrl_id_2, step_val,
+            p_end
+        );
+    }
+    
+    if (max_param.param_type == MaxParam::IntNumber)
+    {
+        int min_val = -10000;
+        int max_val = 10000;
+        int step_val = 1;
+        int def_val = 50;
+
+        if (osl_param.hasMin)
+            min_val = static_cast<int>(osl_param.minValue);
+
+        if (osl_param.hasMax)
+            max_val = static_cast<int>(osl_param.maxValue);
+
+        if (osl_param.hasDefault)
+            def_val = static_cast<int>(osl_param.defaultValue.at(0));
+
+        const int ctrl_id_1 = ctrl_id++;
+        const int ctrl_id_2 = ctrl_id++;
+
+        pb_desc->AddParam(
+            param_id++,
+            param_str.c_str(),
+            TYPE_INT,
+            P_ANIMATABLE,
+            string_id,
+            p_default, def_val,
+            p_range, min_val, max_val,
+            p_ui, TYPE_SPINNER, EDITTYPE_INT, ctrl_id_1, ctrl_id_2, step_val,
+            p_end
+        );
+    }
+
+    if (max_param.param_type == MaxParam::IntCheckbox)
+    {
+        int def_val = TRUE;
+        if (osl_param.hasDefault)
+            def_val = static_cast<int>(osl_param.defaultValue.at(0));
+
+        const int ctrl_id_1 = ctrl_id++;
+
+        pb_desc->AddParam(
+            param_id++,
+            param_str.c_str(),
+            TYPE_INT,
+            0,
+            string_id,
+            p_default, def_val,
+            p_ui, TYPE_SINGLECHEKBOX, ctrl_id_1,
+            p_end
+        );
+    }
+    
+    if (max_param.param_type == MaxParam::IntMapper)
+    {
+        int def_val = 0;
+        if (osl_param.hasDefault)
+            def_val = static_cast<int>(osl_param.defaultValue.at(0));
+
+        const int ctrl_id_1 = ctrl_id++;
+        const int p_id = param_id++;
+        
+        pb_desc->AddParam(
+            p_id,
+            param_str.c_str(),
+            TYPE_INT,
+            0,
+            string_id,
+            p_ui, TYPE_INT_COMBOBOX, ctrl_id_1,
+            0,
+            p_end
+        );
+        
+        std::vector<std::string> fields;
+        asf::tokenize(osl_param.options, "|", fields);
+        const int num_items = static_cast<int>(fields.size());
+
+        Tab<int> string_items;
+        Tab<int> string_item_values;
+        string_items.SetCount(num_items);
+        string_item_values.SetCount(num_items);
+        std::vector<std::string> subfields;
+        for (size_t i = 0, e = fields.size(); i < e; ++i)
```

Comment:
Since we're in C++11 you can start using range-based for loops:

    for (const std::string& field : fields)

You don't necessarily need to change this code, but for future code it makes sense to use the simpler notation.

Reply:
I think `i` as and index variable is required here..

Reply:
You're right.